Add functions to update subgroups in user group objects and `is_user_in_group` function.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
Commit summary -
- Add `subgroups` field to user group objects.
- Add functions to update subgroups in user group objects on receiving events.
- Rename `is_member_of` function to `is_direct_member_of`.
- Add `get_recursives_subgroups` function.
- Add `is_user_in_group` function

Fixes part of: <!-- Issue link, or clear description.--> #19525 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on static/js/user_groups.ts:
```diff
@@ -153,3 +153,26 @@ export function get_user_groups_of_user(user_id: number): UserGroup[] {
+    const subgroup_ids = new Set(user_group.subgroups);
+
+    for (const subgroup_id of subgroup_ids) {
+        const user_group = user_group_by_id_dict.get(subgroup_id);
```
We do not need to use another `groups_checked` like set here since `subgroup_ids` is already a set and will not allow adding an already existing element. (Just mentioning here for clarity.)

timabbott replied:
I think the `groups_checked` set might be helpful for reasons noted in my comment above -- making the code more obviously correct.

timabbott commented on static/js/user_groups.ts:
```diff
@@ -8,6 +8,7 @@ type UserGroup = {
     name: string;
     members: Set<number>;
     is_system_group: boolean;
+    subgroups: Set<number>;
```
Do we want to call this `direct_subgroups` for clarity? It seems like it might be helpful given that some functions will want to look at the full recursively computed set of subgroups of groups.

sahil839 replied:
Yes, we can do that. In that case do we want same in the user group objects sent from server too?

sahil839 replied:
It would be better to do change in the backend at this point of time as clients do not use it currently, so no issue of backwards compatibility and I observed that would also result in duplicate code in tests if server returns `subgroups` field and we use `direct_subgroups` here.

timabbott commented on static/js/user_groups.ts:
```diff
@@ -153,3 +153,26 @@ export function get_user_groups_of_user(user_id: number): UserGroup[] {
+        }
+
+        for (const group_id of user_group.subgroups) {
+            subgroup_ids.add(group_id);
```
So, the semantics of adding something to a `set` while iterating through it are not always well-defined in programming languages; do you know how JavaScript handles that case? I would prefer a design that is correct regardless of how browsers implement that detail, but barring that, a comment making clear it's correct.

sahil839 replied:
I tried some print debugging and it seems that the added element is visited in the loop as expected. The [documentation here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach) mentions that using `forEach` will gurantee that the added element will be visited. I cannot find any robust source for `for..of` syntax that I have used here but a pretty old [Stackoverflow thread](https://stackoverflow.com/questions/28306756/is-it-safe-to-delete-elements-in-a-set-while-iterating-with-for-of#:~:text=Yes%2C%20it%20is%20perfectly%20fine,a%20set%20while%20iterating%20it.) mentions that we can do this.

timabbott replied:
The point here is about readability, not whether it works -- if it requires that kind of googling to determine that it's correct, then folks reading this will worry whether it is actually correct. I'll think about whether we want a comment or to actually change it.

timabbott replied:
I added a comment.

timabbott commented on static/js/user_groups.ts:
```diff
@@ -153,3 +153,26 @@ export function get_user_groups_of_user(user_id: number): UserGroup[] {
+            return undefined;
+        }
+
+        for (const group_id of user_group.subgroups) {
```
I don't like the variable names of `user_group_id`, `subgroup_id`, `group_id` here.  I think best would be:
* `direct_subgroup_id` for this loop variable.
* `subgroup_id` for the outer loop variable is probably fine, but then the `const user_group` variable should be `const subgroup`, I think.
* `user_group_id` is OK at the top, but maybe something like `target_group_id` might feel better.

timabbott commented on frontend_tests/node_tests/user_groups.js:
```diff
@@ -127,3 +127,53 @@ run_test("user_groups", () => {
+    assert.deepEqual(user_groups.get_recursive_subgroups(admins.id), new Set([4]));
+    assert.deepEqual(user_groups.get_recursive_subgroups(all.id), new Set([4, 1, 2, 3]));
+    assert.deepEqual(user_groups.get_recursive_subgroups(test.id), new Set([2, 4, 1, 3]));
+    assert.deepEqual(user_groups.get_recursive_subgroups(foo.id), new Set([]));
```
This test set has an arrangement that I think we don't expect to happen in reality, which is groups containing their a group that contains them. I think that's OK for the tests of this logic, but deserves a comment. Perhaps just above this block:

```
This test setup has a state that won't appear in real data: Groups 2 and 3 each contain the other. We test this corner case because it is a simple way to verify whether our algorithm correctly avoids visiting groups multiple times when determining recursive subgroups. A test case that can occur in practice and would be problematic without this optimization is a tree where each layer connects to every node in the next layer.
```

timabbott commented on static/js/user_groups.ts:
```diff
@@ -176,3 +176,26 @@ export function get_recursive_subgroups(user_group_id: number): Set<number> | un
+        blueslip.error(`Could not find user group with ID ${user_group_id}`);
+        return false;
+    }
+    if (is_direct_member_of(user_group_id, user_id)) {
```
Looking at this, it really feels like the parameters for `is_direct_member_of` are backwards -- `user_id` is a member of `user_group_id`, not the other way around. Do you think it's worth doing another prep commit to swap the parameter order for this function?

timabbott commented on static/js/user_groups.ts:
```diff
@@ -176,3 +176,26 @@ export function get_recursive_subgroups(user_group_id: number): Set<number> | un
+    }
+
+    const subgroup_ids = get_recursive_subgroups(user_group_id);
+    if (subgroup_ids === undefined) {
```
Under what circumstance do we expect it to return `undefined`? I would sorta hope that once we've successfully fetched the `user_group` object above, that it's not possible to fail in that way. If it's just the possibility that the user group ID itself was invalid, which we've already checked above, one idea would be to have `get_recursive_subgroups` accept a group rather than and ID as its parameter. (We can add a `get_recursive_subgroups_by_id` if we need to accept an ID in some future code path).

sahil839 replied:
`get_recursive_subgroups` also returns `undefined` if we cannot get group object of any one of the subgroup id (we need the object to get further subgroups of that group).

I also thought of passing directly group instead of id to `get_recursive_subgroups` but thought we might need to accept ID in future. But adding separate function to accept ID in future if required seems nice idea.

timabbott replied:
Cool. I think this is probably fine to do as a follow-up if we can make it work. 

We will need to consider the possibility that in some situations, we will not have access to subgroups of a group, if we decide to add support for limiting access to viewing a group exists or its membership. Not something to worry about for now, I think, since it's possible we won't go for that type of feature.

sahil839 commented:
Updated the PR with following changes -
- Added commit to swap parameter order in `is_direct_member_of`.
- Updated variable names in `get_recursive_subgroups`.
- Added comment in the tests for the case that we don't expect to happen in reality.

Also, added comments for the other questions above (linking them below)-
- https://github.com/zulip/zulip/pull/21997#discussion_r865101382
- https://github.com/zulip/zulip/pull/21997#discussion_r865147382
- https://github.com/zulip/zulip/pull/21997#discussion_r865150137

timabbott commented:
Merged, after adding a comment as noted above. I think a few of the threads still have outstanding comments that could be useful follow-ups.
